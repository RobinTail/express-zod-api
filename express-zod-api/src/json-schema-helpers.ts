import type { JSONSchema } from "@zod/core";
import { combinations, isObject } from "./common-helpers";

const isJsonObjectSchema = (
  subject: JSONSchema.BaseSchema,
): subject is JSONSchema.ObjectSchema => subject.type === "object";

export const flattenIO = (jsonSchema: JSONSchema.BaseSchema) => {
  const stack = [{ entry: jsonSchema, isOptional: false }];
  const flat: Required<
    Pick<
      JSONSchema.ObjectSchema,
      "type" | "properties" | "required" | "examples"
    >
  > = {
    type: "object",
    properties: {},
    required: [],
    examples: [],
  };
  while (stack.length) {
    const { entry, isOptional } = stack.shift()!;
    if (entry.allOf)
      stack.push(...entry.allOf.map((one) => ({ entry: one, isOptional })));
    if (entry.anyOf) {
      stack.push(
        ...entry.anyOf.map((one) => ({ entry: one, isOptional: true })),
      );
    }
    if (entry.oneOf) {
      stack.push(
        ...entry.oneOf.map((one) => ({ entry: one, isOptional: true })),
      );
    }
    if (!isJsonObjectSchema(entry)) continue;
    if (entry.properties) {
      Object.assign(flat.properties, entry.properties);
      if (!isOptional && entry.required) flat.required.push(...entry.required);
    }
    if (entry.examples) {
      if (isOptional) {
        flat.examples.push(...entry.examples);
      } else {
        flat.examples = combinations(
          flat.examples.filter(isObject),
          entry.examples.filter(isObject),
          ([a, b]) => ({ ...a, ...b }),
        );
      }
    }
    if (entry.propertyNames) {
      const keys: string[] = [];
      if (typeof entry.propertyNames.const === "string")
        keys.push(entry.propertyNames.const);
      if (entry.propertyNames.enum) {
        keys.push(
          ...entry.propertyNames.enum.filter((one) => typeof one === "string"),
        );
      }
      const value = { ...Object(entry.additionalProperties) }; // it can be bool
      for (const key of keys) flat.properties[key] = value;
      if (!isOptional) flat.required.push(...keys);
    }
  }
  if (flat.required.length > 1) flat.required = [...new Set(flat.required)]; // drop duplicates
  return flat;
};
