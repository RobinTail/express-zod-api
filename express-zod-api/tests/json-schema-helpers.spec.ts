import { z } from "zod";
import { flattenIO } from "../src/json-schema-helpers.ts";

describe("JSON Schema helpers", () => {
  describe("flattenIO()", () => {
    test("should pass the object schema through", () => {
      const subject = flattenIO({
        type: "object",
        properties: { one: { type: "string" } },
        required: ["one"],
        examples: [{ one: "test" }],
      });
      expect(subject).toMatchSnapshot();
    });

    test("should return object schema for the union of object schemas", () => {
      const subject = flattenIO({
        oneOf: [
          {
            type: "object",
            properties: { one: { type: "string" } },
            required: ["one"],
            examples: [{ one: "test" }],
          },
          {
            type: "object",
            properties: { two: { type: "number" } },
            required: ["two"],
            examples: [{ two: "jest" }],
          },
        ],
      });
      expect(subject).toMatchSnapshot();
    });

    test("should return object schema for the intersection of object schemas", () => {
      const subject = flattenIO({
        allOf: [
          {
            type: "object",
            properties: { one: { type: "string" } },
            required: ["one"],
            examples: [{ one: "test" }],
          },
          {
            type: "object",
            properties: { two: { type: "number" } },
            required: ["two"],
            examples: [{ two: "jest" }],
          },
        ],
      });
      expect(subject).toMatchSnapshot();
    });

    test("should use top level examples of the intersection", () => {
      const subject = flattenIO({
        examples: [{ one: "test", two: "jest" }],
        allOf: [
          {
            type: "object",
            properties: { one: { type: "string" } },
            required: ["one"],
          },
          {
            type: "object",
            properties: { two: { type: "number" } },
            required: ["two"],
          },
        ],
      });
      expect(subject).toMatchSnapshot();
    });

    test("should pull examples up from object schema props", () => {
      const subject = flattenIO({
        allOf: [
          {
            type: "object",
            properties: { one: { type: "string", examples: ["test", "jest"] } },
            required: ["one"],
          },
          {
            type: "object",
            properties: { two: { type: "number", examples: [123] } },
            required: ["two"],
          },
        ],
      });
      expect(subject).toMatchSnapshot();
    });

    test("should handle records", () => {
      const subject = z.toJSONSchema(
        z
          .record(z.literal(["one", "two"]), z.string())
          .meta({
            examples: [
              { one: "test", two: "jest" },
              { one: "some", two: "another" },
            ],
          })
          .or(
            z
              .record(z.enum(["three", "four"]), z.number())
              .meta({ examples: [{ three: 123, four: 456 }] }),
          ),
      );
      expect(flattenIO(subject)).toMatchSnapshot();
    });
  });
});
