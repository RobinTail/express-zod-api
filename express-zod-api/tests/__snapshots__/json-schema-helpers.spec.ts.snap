// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`JSON Schema helpers > flattenIO() > should handle records 1`] = `
{
  "examples": [
    {
      "one": "test",
      "two": "jest",
    },
    {
      "one": "some",
      "two": "another",
    },
    {
      "four": 456,
      "three": 123,
    },
  ],
  "properties": {
    "four": {
      "type": "number",
    },
    "one": {
      "type": "string",
    },
    "three": {
      "type": "number",
    },
    "two": {
      "type": "string",
    },
  },
  "type": "object",
}
`;

exports[`JSON Schema helpers > flattenIO() > should pass the object schema through 1`] = `
{
  "examples": [
    {
      "one": "test",
    },
  ],
  "properties": {
    "one": {
      "type": "string",
    },
  },
  "required": [
    "one",
  ],
  "type": "object",
}
`;

exports[`JSON Schema helpers > flattenIO() > should return object schema for the intersection of object schemas 1`] = `
{
  "examples": [
    {
      "one": "test",
      "two": "jest",
    },
  ],
  "properties": {
    "one": {
      "type": "string",
    },
    "two": {
      "type": "number",
    },
  },
  "required": [
    "one",
    "two",
  ],
  "type": "object",
}
`;

exports[`JSON Schema helpers > flattenIO() > should return object schema for the union of object schemas 1`] = `
{
  "examples": [
    {
      "one": "test",
    },
    {
      "two": "jest",
    },
  ],
  "properties": {
    "one": {
      "type": "string",
    },
    "two": {
      "type": "number",
    },
  },
  "type": "object",
}
`;

exports[`JSON Schema helpers > flattenIO() > should use top level examples of the intersection 1`] = `
{
  "examples": [
    {
      "one": "test",
      "two": "jest",
    },
  ],
  "properties": {
    "one": {
      "type": "string",
    },
    "two": {
      "type": "number",
    },
  },
  "required": [
    "one",
    "two",
  ],
  "type": "object",
}
`;
